### PR TITLE
[video_player] Feature/add video caching feature

### PR DESCRIPTION
**Pull request by TimHoogstrate**
This resolves feature issue: https://github.com/flutter/flutter/issues/28094

Added caching functionality to video_player, video_player_android, video_player_avfoundation. Changed the platform interface.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart [hunk 1]
```diff
@@ -143,6 +156,24 @@ class DataSource {
...
+  ///
+  /// This will be in different formats depending on the [DataSourceType] of
+  /// the original video.
+  final int? maxFileSize;
+
+  /// The URI to the video file.
+  ///
+  /// This will be in different formats depending on the [DataSourceType] of
+  /// the original video.
+  final bool? enableCache;
```
Comment (TimHoogstrate) [inline packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:175]: add proper documentation

File: packages/video_player/video_player_avfoundation/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,10 @@
 ## NEXT
 
+* Adds caching functionality for network video.
 * Updates minimum supported SDK version to Flutter 3.3/Dart 2.18.
+* Adds caching functionality for network video.
```
Comment (TimHoogstrate) [inline packages/video_player/video_player_avfoundation/CHANGELOG.md:5]: merge glitch

**Comment by tarrinneal** [comment]
Hey @TimHoogstrate it seems like there is an "unknown" author that is connected to this pr, I'm not sure if this is a bug or if someone else contributed, but it is causing the cla check to fail. 

We aren't allowed to begin our review process until the cla check is passing. 

Are you the only contributor to the pr?

**Comment by stuartmorgan** [comment]
> @stuartmorgan is this potentially an issue with the cla bot? I don't see any other contributors listed in the commit history.

It's possible, but every time I've seen something like this before it's because there's some commit in the chain that used the other email address. I can't audit it from the web UI; does a local `git log` with author filtering turn anything up?

**Comment by tarrinneal** [comment]
> > @stuartmorgan is this potentially an issue with the cla bot? I don't see any other contributors listed in the commit history.
> 
> It's possible, but every time I've seen something like this before it's because there's some commit in the chain that used the other email address. I can't audit it from the web UI; does a local `git log` with author filtering turn anything up?

The strange thing is there isn't an email it's just listed as `Author: @UNKNOWN_LOGIN <UNKNOWN_EMAIL>` on the cla bot.

**Comment by tarrinneal** [comment]
> > @stuartmorgan is this potentially an issue with the cla bot? I don't see any other contributors listed in the commit history.
> 
> It's possible, but every time I've seen something like this before it's because there's some commit in the chain that used the other email address. I can't audit it from the web UI; does a local `git log` with author filtering turn anything up?

Nothing shows up when I filter out @TimHoogstrate in the git log

**Comment by ditman** [comment]
The problem with the CLA is that @TimHoogstrate probably has his local Git client configuration a little bit messed up. 

If you look at one of the raw commits in this PR, like: https://github.com/flutter/packages/commit/6150d7c2528f2c2764dbf8f05811f3f4ce8fccab.patch (note the extra `.patch` at the end), you'll see that the author (`From`) field is:

```
From: tim Hoogstrate <>
```

That should contain at least Tim's commit email address (the username is not that relevant, github for example will always show your github username, and not what you've setup locally, if your commit email address matches).

@TimHoogstrate you should configure your local git client with your username and commit email address, as described in points ❷ and ❸ here (especially ❸!):

* https://docs.github.com/en/get-started/quickstart/set-up-git#setting-up-git

(And possibly squash the commits in this branch so the new configuration takes effect here and this can be landed!)

File: packages/video_player/video_player_web/CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,6 @@
 ## NEXT
 
+* Adds caching functionality for network video.
```
Comment (ditman) [inline packages/video_player/video_player_web/CHANGELOG.md:3]: No need to modify anything in `video_player_web` for this feature, right? Unless you're going to implement the new methods that were added in the platform interface as NOOPs so they don't crash.

If this is only going to be metadata changes for the web, you should be able to revert them!

File: packages/video_player/video_player_avfoundation/example/ios/Runner.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -582,18 +582,22 @@
 			baseConfigurationReference = 9740EEB21CF90195004384FC /* Debug.xcconfig */;
 			buildSettings = {
 				ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon;
+				DEVELOPMENT_TEAM = 7FTC5HJTQN;
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/example/ios/Runner.xcodeproj/project.pbxproj:585]: don't check this in

File: packages/video_player/video_player_avfoundation/example/ios/Runner.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -582,18 +582,22 @@
...
+				LD_RUNPATH_SEARCH_PATHS = (
+					"$(inherited)",
+					"@executable_path/Frameworks",
+				);
 				LIBRARY_SEARCH_PATHS = (
 					"$(inherited)",
 					"$(PROJECT_DIR)/Flutter",
 				);
-				PRODUCT_BUNDLE_IDENTIFIER = dev.flutter.videoPlayerExample;
+				PRODUCT_BUNDLE_IDENTIFIER = dev.flutter.videoPlayerExample4;
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/example/ios/Runner.xcodeproj/project.pbxproj:600]: changes in this file seems unintentional?

File: packages/video_player/video_player_avfoundation/example/ios/Runner/Info.plist [hunk 6]
```diff
@@ -48,9 +52,5 @@
 		<string>UIInterfaceOrientationLandscapeLeft</string>
 		<string>UIInterfaceOrientationLandscapeRight</string>
 	</array>
-	<key>CADisableMinimumFrameDurationOnPhone</key>
-	<true/>
-	<key>UIApplicationSupportsIndirectInputEvents</key>
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/example/ios/Runner/Info.plist:53]: accidental change?

File: packages/video_player/video_player_avfoundation/example/ios/RunnerTests/Info.plist [hunk 7]
```diff
@@ -18,5 +18,10 @@
 	<string>1.0</string>
 	<key>CFBundleVersion</key>
 	<string>1</string>
+    <key>NSAppTransportSecurity</key>
+    <dict>
+    <key>NSAllowsArbitraryLoads</key>
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/example/ios/RunnerTests/Info.plist:23]: is this required? if so, may have to inform developers about this change

File: packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m [hunk 8]
```diff
@@ -126,6 +126,7 @@ - (void)testBlankVideoBugWithEncryptedVideoStreamAndInvertedAspectRatioBugForSom
                 uri:@"https://flutter.github.io/assets-for-api-docs/assets/videos/bee.mp4"
         packageName:nil
          formatHint:nil
+              enableCache:nil
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m:129]: whats the diff between nil and false?

Comment (hellohuanlin) [reply packages/video_player/video_player_avfoundation/example/ios/RunnerTests/VideoPlayerTests.m:129]: also nit indentation

File: packages/video_player/video_player_avfoundation/example/lib/mini_controller.dart [hunk 9]
```diff
@@ -349,11 +358,25 @@ class MiniController extends ValueNotifier<VideoPlayerValue> {
     }
   }
 
+  // Clears the media cache
+  void _applyClearCache() {
+    _platform.clearCache(textureId);
+  }
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/example/lib/mini_controller.dart:364]: can just be 

```
void clearCache() {
    _platform.clearCache(textureId);
}
```

File: packages/video_player/video_player_avfoundation/example/lib/mini_controller.dart [hunk 10]
```diff
@@ -349,11 +358,25 @@ class MiniController extends ValueNotifier<VideoPlayerValue> {
...
+
   /// The position in the current video.
   Future<Duration?> get position async {
     return _platform.getPosition(_textureId);
   }
 
+  /// Returns if cache is supported for network media.
+  Future<bool?> isCacheSupportedForNetworkMedia(String url) async {
+    return _applyIsCacheSupported(url);
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/example/lib/mini_controller.dart:373]: what does nil return mean?

File: packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.h [hunk 11]
```diff
@@ -0,0 +1,12 @@
+#import <Foundation/Foundation.h>
+
+typedef NS_ENUM(NSUInteger, CacheActionType) { CacheActionTypeLocal = 0, CacheActionTypeRemote };
+
+@interface CacheAction : NSObject
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.h:5]: need prefix per objc convention

File: packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.h [hunk 12]
```diff
@@ -0,0 +1,12 @@
+#import <Foundation/Foundation.h>
+
+typedef NS_ENUM(NSUInteger, CacheActionType) { CacheActionTypeLocal = 0, CacheActionTypeRemote };
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.h:3]: local and remote dont sound like "actions". maybe just cache type? also what does remote cache mean?

File: packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.m [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+  if (!NSEqualRanges(object.range, self.range)) {
+    return NO;
+  }
+
+  if (object.actionType != self.actionType) {
+    return NO;
+  }
+
+  return YES;
+}
```
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.m:24]: can just be `return range equal && action type equal`

**Review by hellohuanlin — CHANGES_REQUESTED** [review]
I reviewed some of the iOS part and stopped half way since there are plenty of comments i made. I can take a closer look after you address those.

File: packages/video_player/video_player_web/CHANGELOG.md (hunk 3 above)
Comment (TimHoogstrate) [reply packages/video_player/video_player_web/CHANGELOG.md:3]: You are correct, it seems that, and I have tested this on different browsers, caching is by default supported (probably based on browser settings off course. This is also what I was expecting. Therefore we don't have to do anything yet for web. I'll cleanup some web changes.

File: packages/video_player/video_player_avfoundation/example/ios/Runner.xcodeproj/project.pbxproj (hunk 4 above)
Comment (TimHoogstrate) [reply packages/video_player/video_player_avfoundation/example/ios/Runner.xcodeproj/project.pbxproj:585]: True, I'll revert them.

File: packages/video_player/video_player_avfoundation/example/ios/Runner.xcodeproj/project.pbxproj (hunk 5 above)
Comment (TimHoogstrate) [reply packages/video_player/video_player_avfoundation/example/ios/Runner.xcodeproj/project.pbxproj:600]: true, thanks for finding this :)

File: packages/video_player/video_player_avfoundation/example/ios/RunnerTests/Info.plist (hunk 7 above)
Comment (TimHoogstrate) [reply packages/video_player/video_player_avfoundation/example/ios/RunnerTests/Info.plist:23]: No it was for testing purposes of different free video files. I think I can remove them.

File: packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.h (hunk 12 above)
Comment (TimHoogstrate) [reply packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.h:3]: Remote and local cache types are used for optimisation purposes. You basically can use parts of existing local cache's regardless of the validity. From objective-c documentation: `No
    existing local cache data, regardless of its freshness or validity,
    should be used to satisfy a URL load request.`

Comment (hellohuanlin) [reply packages/video_player/video_player_avfoundation/ios/Classes/Cache/Cache/CacheAction.h]: i meant what are local cache and remote cache? can you link me to any documentation?

File: packages/video_player/video_player_web/CHANGELOG.md (hunk 3 above)
Comment (ditman) [reply packages/video_player/video_player_web/CHANGELOG.md]: Yes, caching is on by default in all web browsers. Users can tweak its behavior a little bit. Server response headers can be used to control its behavior (like [`cache-control: no-cache`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#no-cache)). In general it is *hard* to change the default (in the browser, you'd need a special service worker to override the default behavior). Thanks for cleaning this up!

**Comment by TimHoogstrate** [comment]
Made a new PR with only one commit, hope it resolves the contribution issue. This PR became a mess.

**Pull request closed without merging**